sycl/include/sycl/ext/oneapi/matrix/matrix-unified.hpp

```diff
@@ -0,0 +1,221 @@
+//===------- matrix-unified.hpp - SYCL matrix extension ----*- C++ -*------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// ===--------------------------------------------------------------------=== //
+
+#pragma once
+#include <sycl/ext/oneapi/matrix/matrix-tensorcores.hpp>
+
+namespace sycl {
+__SYCL_INLINE_VER_NAMESPACE(_V1) {
+namespace ext {
+namespace oneapi {
+namespace experimental {
+namespace matrix {
+
+template <typename Group, typename T, use Use, size_t Rows, size_t Cols,
+          layout Layout>
+struct joint_matrix {
+
+#if defined(__SYCL_DEVICE_ONLY__) && defined(__SPIR__)
+  // TODO: Intel case here: we use the ext_oneapi_cuda case also for the host,
+  // because the Intel SPIRV functions will not be host compilable.
+#else
+  sycl::ext::oneapi::detail::joint_matrix_cuda<T, Use, Rows, Cols, Layout>
+      cuda_impl;
+#endif // defined(__SYCL_DEVICE_ONLY__) && defined(__SPIR__)
+
+  joint_matrix() {
+#ifndef __SYCL_DEVICE_ONLY__
+    throw runtime_error("joint matrix is not supported on host device.",
+                        PI_ERROR_INVALID_DEVICE);
+#endif
+  }
+};
+
+template <typename Group, typename T, use Use, size_t Rows, size_t Cols,
+          layout Layout>
+inline __SYCL_ALWAYS_INLINE wi_data<Group, T, Use, Rows, Cols, Layout>
+get_wi_data(Group sg, joint_matrix<Group, T, Use, Rows, Cols, Layout> &jm) {
+#if defined(__SYCL_DEVICE_ONLY__)
+#if defined(__NVPTX__)
+  std::ignore = sg;
+  return wi_data(jm);
+#else
+  // TODO add Intel impl.
```

Comment:
for intel, we have to provide another class wi_data_intel, so the return type should be "decltype(auto)" since we can't provide wi_data in both cuda&intel's header.

If the return type is "decltype(auto)", for host, we have to return something.
so, could we switch back to your previous code of get_wi_data but we still keep it a free function?